Mathlib/CategoryTheory/MorphismProperty/Tactic.lean

```diff
@@ -0,0 +1,231 @@
+        let fvars := (← Term.collectUnassignedMVars proofTerm) ++ fvars
+        let proofTerm ← instantiateMVars (← Term.levelMVarToParam
+          (← mkLambdaFVars (usedOnly := true) fvars proofTerm))
+        let statement ← instantiateMVars (← mkForallFVars (usedOnly := true) fvars statement)
+        let decl : TheoremVal :=
+        { name := name
+          levelParams := (collectLevelParams {} proofTerm).params.toList
+          type := statement
+          value := proofTerm }
+        -- We add the declaration and add it as an instance
+        addAndCompile (.thmDecl decl)
+        addInstance name .global config.priority
+        logMsg := logMsg ++ m!"Added instance {name}:{indentExpr statement}\n\n"
+        success := success + 1
+      catch
+      | err => logMsg :=
+        logMsg ++ m!"Failed to add instance for {lemmaName}:\n{err.toMessageData}\n\n"
+    if success == 0 then
+      throwError m!"Failed to add any instances:\n\n" ++ logMsg
+    if verbose then
+      logMsg := m!"Successfully added {success} instances out of {lemmas.size}.\n\n" ++ logMsg
+      logInfo logMsg
+  else
+    throwError m!"Failed to generate lemmas. {classExpr} is not a class."
+
+/--
+`addMorphismPropertyInstances P` is a command that
+generates instance lemmas for `P : MorphismProperty _`.
+
+It runs through all the lemmas tagged with `morphismPropertyInstance` and attempt to
+generate a lemma with the morphism property in the lemma substituted with `P`, provided that
+the appropriate stability instances on the morphism property are present.
+
+For example, given
+```
+@[morphismPropertyInstance]
+lemma foo {P : MorphismProperty C} [P.ContainsIdentities] {X} : P (𝟙 X)
+```
+Then the following command
+```
+addMorphismPropertyInstances Bar
+```
+would generate the instance
+```
+instance {X} : Bar (𝟙 X)
+```
+if `Bar` is a class and a `Bar.ContainsIdentities` instance is available.
+
+Use `addMorphismPropertyInstances?` to print the list of generated lemmas.
+-/
+elab "addMorphismPropertyInstances" classTerm:term : command =>
+  addMorphismPropertyInstancesAux classTerm
+
+/--
+`addMorphismPropertyInstances? P` is the verbose version of `addMorphismPropertyInstances P`,
+which is a command that generates instance lemmas for `P : MorphismProperty _`.
+
+It runs through all the lemmas tagged with `morphismPropertyInstance` and attempt to
+generate a lemma with the morphism property in the lemma substituted with `P`, provided that
+the appropriate stability instances on the morphism property are present.
+
+For example, given
+```
+@[morphismPropertyInstance]
+lemma foo {P : MorphismProperty C} [P.ContainsIdentities] {X} : P (𝟙 X)
+```
+Then the following command
+```
+addMorphismPropertyInstances Bar
+```
+would generate the instance
+```
+instance {X} : Bar (𝟙 X)
+```
+if `Bar` is a class and a `Bar.ContainsIdentities` instance is available.
+-/
+elab "addMorphismPropertyInstances?" classTerm:term : command =>
```

Comment:
can this use inherit doc or something to not repeat the doc

Reply:
All the docstrings are slightly different so I think it might not be possible to do so.

Reply:
I guess I'm suggesting to combine the ones that are basically the same (just explaining the ? variant is the difference). This would be slightly better for maintenance going forward but if your happy with it as is I don't mind